File: javascript/test/parser.test.js
import {readFileSync} from 'fs'; 
import {strict as assert} from 'assert';
import {FountainParser} from "../src/parser.js";

describe('FountainParser', () => {

  describe('Scratch', () => {
    it('should match', () => {

      const source = readFileSync('../tests/Scratch.fountain', 'utf-8');
      const match = readFileSync('../tests/Scratch.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });

  describe('SceneHeading', () => {
    it('should match', () => {

      const source = readFileSync('../tests/SceneHeading.fountain', 'utf-8');
      const match = readFileSync('../tests/SceneHeading.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });

  describe('Actions', () => {
    it('should match merged', () => {

      const source = readFileSync('../tests/Action.fountain', 'utf-8');
      const match = readFileSync('../tests/Action-Merged.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });

    it('should match unmerged', () => {

      const source = readFileSync('../tests/Action.fountain', 'utf-8');
      const match = readFileSync('../tests/Action-Unmerged.txt', 'utf-8');

      let fp = new FountainParser();
      fp.mergeActions = false;
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });

  });

  describe('Character', () => {
    it('should match', () => {

      const source = readFileSync('../tests/Character.fountain', 'utf-8');
      const match = readFileSync('../tests/Character.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });

  describe('Dialogue', () => {
    it('should match merged', () => {

      const source = readFileSync('../tests/Dialogue.fountain', 'utf-8');
      const match = readFileSync('../tests/Dialogue-Merged.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });

    it('should match unmerged', () => {

      const source = readFileSync('../tests/Dialogue.fountain', 'utf-8');
      const match = readFileSync('../tests/Dialogue-Unmerged.txt', 'utf-8');

      let fp = new FountainParser();
      fp.mergeDialogue = false;
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });

  describe('Parenthetical', () => {
    it('should match', () => {

      const source = readFileSync('../tests/Parenthetical.fountain', 'utf-8');
      const match = readFileSync('../tests/Parenthetical.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });

  describe('Lyrics', () => {
    it('should match', () => {

      const source = readFileSync('../tests/Lyrics.fountain', 'utf-8');
      const match = readFileSync('../tests/Lyrics.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });

  describe('Transition', () => {
    it('should match', () => {

      const source = readFileSync('../tests/Transition.fountain', 'utf-8');
      const match = readFileSync('../tests/Transition.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });

  describe('TitlePage', () => {
    it('should match', () => {

      const source = readFileSync('../tests/TitlePage.fountain', 'utf-8');
      const match = readFileSync('../tests/TitlePage.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });

  describe('PageBreak', () => {
    it('should match', () => {

      const source = readFileSync('../tests/PageBreak.fountain', 'utf-8');
      const match = readFileSync('../tests/PageBreak.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });

  describe('LineBreaks', () => {
    it('should match', () => {

      const source = readFileSync('../tests/LineBreaks.fountain', 'utf-8');
      const match = readFileSync('../tests/LineBreaks.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });

  describe('Notes', () => {
    it('should match', () => {

      const source = readFileSync('../tests/Notes.fountain', 'utf-8');
      const match = readFileSync('../tests/Notes.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });

  describe('Boneyards', () => {
    it('should match', () => {

      const source = readFileSync('../tests/Boneyards.fountain', 'utf-8');
      const match = readFileSync('../tests/Boneyards.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });
  
  describe('Sections', () => {
    it('should match', () => {

      const source = readFileSync('../tests/Sections.fountain', 'utf-8');
      const match = readFileSync('../tests/Sections.txt', 'utf-8');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(match, output);
    });
  });

});

File: csharp/FountainTools.Tests/ParserTest.cs
namespace FountainTools.Tests;
using System.IO;
using Fountain;

public class ParserTest
{
    private string loadTestFile(string fileName) {
        return File.ReadAllText("../../../../../tests/"+fileName);
    }

    [Fact]
    public void Scratch()
    {
        string source = loadTestFile("Scratch.fountain");
        string match = loadTestFile("Scratch.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void SceneHeading()
    {
        string source = loadTestFile("SceneHeading.fountain");
        string match = loadTestFile("SceneHeading.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void ActionMerged()
    {
        string source = loadTestFile("Action.fountain");
        string match = loadTestFile("Action-Merged.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void ActionUnmerged()
    {
        string source = loadTestFile("Action.fountain");
        string match = loadTestFile("Action-Unmerged.txt");

        FountainParser fp = new FountainParser();
        fp.MergeActions = false;
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Character()
    {
        string source = loadTestFile("Character.fountain");
        string match = loadTestFile("Character.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void DialogueMerged()
    {
        string source = loadTestFile("Dialogue.fountain");
        string match = loadTestFile("Dialogue-Merged.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void DialogueUnmerged()
    {
        string source = loadTestFile("Dialogue.fountain");
        string match = loadTestFile("Dialogue-Unmerged.txt");

        FountainParser fp = new FountainParser();
        fp.MergeDialogue = false;
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Parenthetical()
    {
        string source = loadTestFile("Parenthetical.fountain");
        string match = loadTestFile("Parenthetical.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Lyrics()
    {
        string source = loadTestFile("Lyrics.fountain");
        string match = loadTestFile("Lyrics.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Transition()
    {
        string source = loadTestFile("Transition.fountain");
        string match = loadTestFile("Transition.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void TitlePage()
    {
        string source = loadTestFile("TitlePage.fountain");
        string match = loadTestFile("TitlePage.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void PageBreak()
    {
        string source = loadTestFile("PageBreak.fountain");
        string match = loadTestFile("PageBreak.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void LineBreaks()
    {
        string source = loadTestFile("LineBreaks.fountain");
        string match = loadTestFile("LineBreaks.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Notes()
    {
        string source = loadTestFile("Notes.fountain");
        string match = loadTestFile("Notes.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Boneyards()
    {
        string source = loadTestFile("Boneyards.fountain");
        string match = loadTestFile("Boneyards.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

   [Fact]
    public void Sections()
    {
        string source = loadTestFile("Sections.fountain");
        string match = loadTestFile("Sections.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

   [Fact]
    public void UTF8()
    {
        string source = loadTestFile("UTF8.fountain");
        string match = loadTestFile("UTF8.txt");

        FountainParser fp = new FountainParser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }
}
